src/nnet3/nnet-component-test.cc

```diff
@@ -54,6 +54,38 @@ void TestNnetComponentAddScale(Component *c) {
   delete c3;
 }
 
+void TestNnetComponentVectorizeUnVectorize(Component *c) {
+  UpdatableComponent *uc = dynamic_cast<UpdatableComponent*>(c);
+  if((uc==NULL) || (uc->NumParameters() == 0))
+    return;
+  UpdatableComponent *uc2 = dynamic_cast<UpdatableComponent*>(uc->Copy());
+  uc2->SetZero(false);
+  Vector<BaseFloat> params(uc2->NumParameters());
+  uc2->Vectorize(&params);
+  KALDI_ASSERT(params.Min()==0.0 && params.Sum()==0.0);
+  uc->Vectorize(&params);
+  uc2->UnVectorize(params);
+  KALDI_ASSERT(uc2->Info() == uc->Info());
+  KALDI_ASSERT(uc2->DotProduct(*uc2) == uc->DotProduct(*uc));
+  delete uc2;
```

Comment:
true, this is a possible error that we need to worry about in that code .
maybe this could be solved by doing 'vectorize' one more time and checking
the result is the same as the first time.

On Wed, Nov 4, 2015 at 4:15 PM, Vijayaditya Peddinti <
notifications@github.com> wrote:

> In src/nnet3/nnet-component-test.cc
> <https://github.com/kaldi-asr/kaldi/pull/307#discussion_r43940500>:
>
> > @@ -54,6 +54,38 @@ void TestNnetComponentAddScale(Component *c) {
> >    delete c3;
> >  }
> >
> > +void TestNnetComponentVectorizeUnVectorize(Component *c) {
> > +  UpdatableComponent *uc = dynamic_cast<UpdatableComponent*>(c);
> > +  if((uc==NULL) || (uc->NumParameters() == 0))
> > +    return;
> > +  UpdatableComponent *uc2 = dynamic_cast<UpdatableComponent*>(uc->Copy());
> > +  uc2->SetZero(false);
> > +  Vector<BaseFloat> params(uc2->NumParameters());
> > +  uc2->Vectorize(&params);
> > +  KALDI_ASSERT(params.Min()==0.0 && params.Sum()==0.0);
> > +  uc->Vectorize(&params);
> > +  uc2->UnVectorize(params);
> > +  KALDI_ASSERT(uc2->Info() == uc->Info());
> > +  KALDI_ASSERT(uc2->DotProduct(*uc2) == uc->DotProduct(*uc));
> > +  delete uc2;
>
> This test wouldn't fail if there is a permutation of rows of the parameter
> matrix.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/307/files#r43940500>.
>